internal/wal/wal_aof.go

```diff
@@ -65,31 +170,212 @@ func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
 	data, err := proto.Marshal(entry)
 	if err != nil {
 		slog.Warn("failed to serialize command", slog.Any("error", err.Error()))
+		return
 	}
 
-	writeBuf.Reset()
-	writeBuf.Grow(4 + len(data))
-	if binary.Write(&writeBuf, binary.BigEndian, uint32(len(data))) != nil {
-		slog.Warn("failed to write entry length to WAL", slog.Any("error", err.Error()))
+	buf := bufferPool.Get().([]byte)
+	buf = buf[:0]
+	defer bufferPool.Put(buf)
+
+	length := uint32(len(data))
+	totalSize := 4 + len(data)
+
+	if cap(buf) < totalSize {
+		buf = make([]byte, totalSize)
+	} else {
+		buf = buf[:totalSize]
+	}
+
+	binary.BigEndian.PutUint32(buf[0:4], length)
+	copy(buf[4:], data)
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	if w.rotationMode == "size" {
+		if err := w.rotateLogIfNeeded(); err != nil {
+			slog.Warn("failed to rotate WAL log", slog.Any("error", err.Error()))
+			return
+		}
 	}
-	writeBuf.Write(data)
 
-	if _, err := w.file.Write(writeBuf.Bytes()); err != nil {
+	if _, err := w.writer.Write(buf); err != nil {
 		slog.Warn("failed to write serialized command to WAL", slog.Any("error", err.Error()))
+		return
 	}
 
-	if err := w.file.Sync(); err != nil {
-		slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+	if w.writeMode == "fsync" {
```

Comment:
The entire writer logic should be moved to a separate `WALWriter` object or a different method. It should expose an interface `Write` which internally can select a buffered write or others depending on the configuration.

Reply:
I am planning to abstract it out based on the below arch. It will also write segments into mmap files if needed.
https://excalidraw.com/#json=5i9YJvPvBYtihEbC4YA_A,fUc_OmTJ8_FyOdh3vd3C4A